Frameworks/include/CFCppBase.h

```diff
@@ -118,15 +103,33 @@ struct CppBase : Details::CppBaseImpl<_Parent> {
         return _id;
     }
 
-    static T* CreateInstance(CFAllocatorRef allocator = kCFAllocatorDefault) {
-        return reinterpret_cast<T*>(
+    template <typename... Args>
+    static T* CreateInstance(CFAllocatorRef allocator, Args&&... args) {
+        static_assert(!std::is_polymorphic<T>::value,
+                      "CoreFoundation::CppBase cannot be used with polymorphic types or classes bearing virtual functions.");
+        T* ret = reinterpret_cast<T*>(
             const_cast<void*>(_CFRuntimeCreateInstance(allocator, GetTypeID(), sizeof(T) - sizeof(__CFRuntimeBase), nullptr)));
-    }
 
-    TImpl _impl;
+        // Technically, accessing any members of __CFRuntimeBase (which are set up by _CFRuntimeCreateInstance) after
+        // new (ret) T() is undefined behaviour, since the constructor chain does not change them from their otherwise
+        // "uninitialized" values.
+        //
+        // To work around that, we save the __CFRuntimeBase at the head of ret, call the constructor, and replace/reinitialize
+        // it. As no constructor has been invoked on ret by the time we save temp, this is not ill-defined.
+        __CFRuntimeBase temp = *reinterpret_cast<__CFRuntimeBase*>(ret);
+
+        new (ret) T(std::forward<Args>(args)...);
+
+        // This, however, may be ill-defined.
```

Comment:
can this be a static_cast? Every T here should work out for that and we should maybe static_assert or enable_if with is_base_of to validate that CreateInstance only gets called on __CFRuntimeBase things. (Would be super nifty to know it was the root base (not a multi inheritance weirdness) which is maybe doable with template magic?)